libs/ledger-live-common/src/families/filecoin/bridge/utils/erc20/tokenAccounts.ts

```diff
@@ -0,0 +1,179 @@
+import contractABI from "./ERC20.json";
+import { RecipientRequired } from "@ledgerhq/errors";
+import { Unit } from "@ledgerhq/types-cryptoassets";
+import { valueFromUnit } from "../../../../../currencies/valueFromUnit";
+
+export const erc20TxnToOperation = (
+  tx: ERC20Transfer,
+  address: string,
+  accountId: string,
+  unit: Unit,
+): Operation[] => {
+  try {
+    const { to, from, timestamp, tx_hash, tx_cid, amount, height } = tx;
+    const value = valueFromUnit(new BigNumber(amount), unit);
+
+    const isSending = address.toLowerCase() === from.toLowerCase();
+    const isReceiving = address.toLowerCase() === to.toLowerCase();
+
+    const fee = new BigNumber(0);
+
+    const date = new Date(timestamp * 1000);
+    const hash = tx_cid ?? tx_hash;
+
+    const ops: Operation[] = [];
+    if (isSending) {
+      ops.push({
+        id: encodeOperationId(accountId, hash, "OUT"),
+        hash,
+        type: "OUT",
+        value: value,
+        fee,
+        blockHeight: height,
+        blockHash: null,
+        accountId,
+        senders: [from],
+        recipients: [to],
+        date,
+        extra: {},
+      });
+    }
+
+    if (isReceiving) {
+      ops.push({
+        id: encodeOperationId(accountId, hash, "IN"),
+        hash,
+        type: "IN",
+        value,
+        fee,
+        blockHeight: height,
+        blockHash: null,
+        accountId,
+        senders: [from],
+        recipients: [to],
+        date,
+        extra: {},
+      });
+    }
+
+    invariant(ops, "filecoin operation is not defined");
+
+    return ops;
+  } catch (e) {
+    log("error", "filecoin error converting erc20 transaction to operation", e);
+    return [];
+  }
+};
+
+export async function buildTokenAccounts(
+  filAddr: string,
+  parentAccountId: string,
+  initialAccount?: Account,
+): Promise<TokenAccount[]> {
+  try {
+    const nativeEthAddr = await convertAddressFilToEth(filAddr);
+    const transfers = await fetchERC20Transactions(nativeEthAddr);
+    const transfersUntangled: { [addr: string]: ERC20Transfer[] } = transfers.reduce(
+      (prev, curr) => {
+        curr.contract_address = curr.contract_address.toLowerCase();
+        if (prev[curr.contract_address]) {
+          prev[curr.contract_address] = [...prev[curr.contract_address], curr];
+        } else {
+          prev[curr.contract_address] = [curr];
+        }
+        return prev;
+      },
+      {},
+    );
+
+    const subs: TokenAccount[] = [];
+    for (const [cAddr, txns] of Object.entries(transfersUntangled)) {
+      const token = findTokenByAddressInCurrency(cAddr, "filecoin");
+      if (!token) {
+        log("error", `filecoin token not found, addr: ${cAddr}`);
+        continue;
+      }
+
+      const balance = await fetchERC20TokenBalance(nativeEthAddr, cAddr);
+      const bnBalance = new BigNumber(balance.toString());
+      const tokenAccountId = encodeTokenAccountId(parentAccountId, token);
+
+      const operations = txns
```

Comment:
Just a question: For the subaccount operations, are all the addresses retrieved from the explorer endpoint expressed in the Ethereum format instead of the Filecoin format?

If a user clicks the "receive" button, does he get an Ethereum format address or a Filecoin format address?

Do you think it is better to convert back to the Filecoin format in the UI? If a user sends ERC20 tokens using his Filecoin account in the UI, I think he expects to see his Filecoin address as the sender in the operation list.

Reply:
> For the subaccount operations, are all the addresses retrieved from the explorer endpoint expressed in the Ethereum format instead of the Filecoin format?

Yes they are, the erc20 endpoints on indexing only stores the address data from the evm world. The translation isn't yet there. 

> If a user clicks the "receive" button, does he get an Ethereum format address or a Filecoin format address?

The user will see a filecoin address format with `f1` protocol. Converting to `f1` -> 0x involves a lookup from node through network call which we think might cause trouble hence the translation isn't supported. More info [here](https://docs.filecoin.io/smart-contracts/filecoin-evm-runtime/address-types#converting-to-a-0x-style-address)

> Do you think it is better to convert back to the Filecoin format in the UI?

It's definitely better user experience IMO. It will definitely cost more network calls because going from 0x to f1 might be tricky, the other way is much easier. We already do convert f1 -> 0x to check if the account is sending or receiving, I think it would be okay to override the sender/reciever address at least when creating operations. What do you think?